sdk/core/Azure.Core/src/Shared/Argument.cs

```diff
@@ -212,5 +212,24 @@ public static string CheckNotNullOrEmpty(string value, string name)
             AssertNotNullOrEmpty(value, name);
             return value;
         }
+
+        /// <summary>
+        /// Throws if <paramref name="value"/> is not null.
+        /// </summary>
+        /// <param name="value">The value to validate.</param>
+        /// <param name="name">The name of the parameter.</param>
+        /// <param name="message">The error message.</param>
+        /// <exception cref="ArgumentException"><paramref name="value"/> is not null.</exception>
+#if AZURE_NULLABLE
+        public static void AssertNull<T>([AllowNull] T value, string name, string message)
+#else
+        public static void AssertNull<T>(T value, string name, string message)
+#endif
+        {
+            if (value is not null)
+            {
+                throw new ArgumentException(message, name);
```

Comment:
We already have an `AssertNotNull` [method above](https://github.com/Azure/azure-sdk-for-net/pull/26654/files#diff-36e10e14f3b4e8d4adc075eab100e86cbc1375c56e0b2228de679aa9d68b48ccR55). Why do we need this method, and why isn't it throwing an `ArgumentNullException`, which is more typical (and extends `ArgumentException`)?

Reply:
We will collapse all the request condition headers(If-Match, if-Not-Match, if-Modified-since, if-not-modified-since) as one parameter RequestConditions, but we need to check if the collapsed parameter 'RequestConditions requestConditions' only set the value to the headers which the service supported. If the value set to unsupported condition header, it will cause unexpected problem. So we need to make sure that the unsupported head field of RequestConditons is not set (that is null). So we add method AssertNull to do this.

Reply:
@heaths, `AssertNotNull` ensures that an argument is not null.  In this case, we want to ensure that an argument *is* null.  Can you help us understand your concern?

Reply:
I misread the doc comment, but my previous question still stands: does every library need this? The shared `Argument` class is partial and has common assertions. You can add your own library-specific ones by declaring your own partial class and still have everything hanging off `Argument`. I did this for Key Vault SDK, for example.

Reply:
@heaths, for code generation, we don't know offhand which libraries will need this -- it's something we would detect in a swagger file and then generate the code for if we identify a case where it's needed.  The advantage of DPG is that we can do this without manual intervention, lowering the cost of development for our team.

Can you help me understand the negative of adding this to the internal `Argument` class?  It sounds like it is a perf concern that now every library will have these methods in shared source?  Is it possible to characterize the actual cost of this, so we can better understand the trade-off we'd be making?

Reply:
Not directly perf, but size. You make a good point since this is for code gen, but I think in the future we need to be more careful. For example, doesn't autorest.csharp already reference a bunch of code gen-related code? You could put it in there. I would consider it, but if possible perhaps in a separate issue/PR. I'll unblock for now, but I think collectively we should try to avoid stuffing too much code into shared code that isn't common enough.